### PR TITLE
Recursive rmi implementation [full ci]

### LINKED ISSUE #2973: Implement `rmi` up to last tag in SPL
We're currently only deleting leaf images when an `rmi` is called on a given `repository:tag`.  Until recently we couldn't have more than 1 tag on a given image ID, but the personality's image cache now supports multiple tags and untagging on `rmi`.  We need to integrate this into the PL delete mechanism to delete upto the next image, thereby actually deleting more than just the last image.

Without this change, pull of an image (tag:repo) which you just deleted will show there's nothing to do because the majority of the data is still in the datastore.  (We picked this to be conservative until the personality's image cache had this tagging/persistance-of-tags which it has now.)

### PR DESCRIPTION
**Pull request by fdawg4l**
Fixes #2973 

Deletes a given image all the way up to the next tag, layer with
degree > 1, or used image.

### CODE REVIEW

File: lib/apiservers/engine/backends/image.go [hunk 1]
```diff
@@ -90,17 +93,54 @@ func (i *Image) ImageDelete(imageRef string, force, prune bool) ([]types.ImageDe
...
+				i, err := imagec.LayerCache().Get(di.ID)
+				if err == nil {
+					id = i.Layer.BlobSum
+				}
+
+				// remove the layer from the layer cache (used by imagec)
+				imagec.LayerCache().Remove(di.ID)
+
+				// form the response
+				imageDeleted := types.ImageDelete{Deleted: strings.TrimPrefix(id, "sha:")}
```
Comment (jzt) [inline lib/apiservers/engine/backends/image.go]: should this be `sha256:` ?

Comment (fdawg4l) [inline lib/apiservers/engine/backends/image.go]: I *just* realized this.  force pushed.

**Comment by jzt** [comment]
LGTM

File: lib/apiservers/engine/backends/image.go [hunk 2]
```diff
@@ -90,17 +93,54 @@ func (i *Image) ImageDelete(imageRef string, force, prune bool) ([]types.ImageDe
...
+		params := storage.NewDeleteImageParamsWithContext(ctx).WithStoreName(storeName).WithID(img.ID).WithKeepNodes(keepNodes)
 		// TODO: This will fail if any containerVMs are referencing the vmdk - vanilla docker
 		// allows the removal of an image (via force flag) even if a container is referencing it
 		// should vic?
-		_, err = PortLayerClient().Storage.DeleteImage(params)
+		res, err := PortLayerClient().Storage.DeleteImage(params)
+
+		// We may have deleted images despite error.  Account for that in the cache.
+		if res != nil {
+			for _, di := range res.Payload {
```
Comment (emlin) [inline lib/apiservers/engine/backends/image.go]: is the di confusing? :)

**Comment by emlin** [comment]
lgtm 👍

File: lib/apiservers/engine/backends/image.go (hunk 2 above)
Comment (fdawg4l) [inline lib/apiservers/engine/backends/image.go]: Yeah, good point.  It meant "deletedImage".  I'll rename it.

File: lib/apiservers/engine/backends/image.go [hunk 3]
```diff
@@ -131,7 +169,7 @@ func (i *Image) ImageDelete(imageRef string, force, prune bool) ([]types.ImageDe
 		// updates
 		refNamed, _ := cache.RepositoryCache().Remove(tags[i], false)
 		dd := types.ImageDelete{Untagged: refNamed}
-		deleted = append(deleted, dd)
+		deletedRes = append(deletedRes, dd)
```
Comment (fdawg4l) [inline lib/apiservers/engine/backends/image.go]: I didn't make this variable but I'll rename it.

**Comment by fdawg4l** [comment]
.

**Pull request merged**